### ISSUE TITLE
Medium.com links parsed as mail links

### ISSUE THREAD

**Issue by gj** [lycheeverse/lychee#161]
E.g.:

```
Errors in new-docs/content/any/project/performance/n_plus_one.md
↯ medium.com/@bretdoucette/n-1-queries-and-guides-avoid-them-a12f02345be5 (Invalid mail address: medium.com/@bretdoucette/n-1-queries-and-guides-avoid-them-a12f02345be5)
```

Presumably this is a bit too eager:

https://github.com/lycheeverse/lychee/blob/cefe38ee25d433dd3b7c61a96ccacbec66aa70a8/src/uri.rs#L59-L61

**Comment by mre** [comment]
> Presumably this is a bit too eager
 
Exactly. Any idea on how to fix that check?
One idea would be to try and parse the part before the `@` as a URL. If that succeeds then assume that the entire link is a URL.

Similar to this:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=aff61feeddf28ae7084cda1015f2ef08

Although in this case the link does not begin with a protocol, so it counts as a relative URL without a base, which cannot be parsed by the `url` crate. (You can check by removing the `https://` in the example above.)

@joesan perhaps you have a better idea?

**Comment by mre** [comment]
In the meantime you could use `--exclude-mail` in case you don't want to check mail addresses anyway. 😅

**Comment by gj** [comment]
> Although in this case the link does not begin with a protocol, so it counts as a relative URL without a base, which cannot be parsed by the `url` crate. (You can check by removing the `https://` in the example above.)

The link actually does begin with `https://` — GitHub Actions UI helpfully (🙄) removes the protocol when formatting URLs.

> Exactly. Any idea on how to fix that check?
> One idea would be to try and parse the part before the `@` as a URL. If that succeeds then assume that the entire link is a URL.

If you only care about URLs that include a protocol, then that might work. But if `example.com` should be treated as a URL, there might be issues since `example.com@example.com` is a valid email address.

It looks like that `url` crate parses `mailto:` links just fine, so I would probably try something like:

1. Try parsing the input string with `Url::parse` (without manually stripping out `mailto:`).
2. If (1) fails, try parsing it as an email.
3. If (1) succeeds and the parsed `scheme` is `"mailto"`, try parsing the parsed `path` as an email, same as (2).
4. If (1) succeeds and the parsed `scheme` is `"http"`/`"https"`, great, we have a URL.
5. If we haven't hit one of the previous cases, fail? Maybe there are other cases to care about — brain's a bit dead at the moment. 😄

**Comment by joesan** [comment]
Would using a Regex to check for E-Mail solve the problem? I mean using Regex to distinguish between E-Mail and URL. So if I can determine using Regex that the given link is an E-Mail, then I'm done. What do you say @mre

**Comment by joesan** [comment]
I was able to use Regex to differentiate E-Mail from URL, but now it panics saying that it cannot find a URL base:

`thread 'extract::test::test_markdown_email_like_link' panicked at 'Expected valid Website URI: RelativeUrlWithoutBase', src/test_utils.rs:40:34`

**Comment by gj** [comment]
> I was able to use Regex to differentiate E-Mail from URL, but now it panics saying that it cannot find a URL base:
> 
> used this as test url: medium.com/@bretdoucette/n-1-queries-and-guides-avoid-them-a12f02345be5
> 
> `thread 'extract::test::test_markdown_email_like_link' panicked at 'Expected valid Website URI: RelativeUrlWithoutBase', src/test_utils.rs:40:34`

That's an error from the `url` crate:

> > Although in this case the link does not begin with a protocol, so it counts as a relative URL without a base, which cannot be parsed by the `url` crate. (You can check by removing the `https://` in the example above.)
> 
> The link actually does begin with `https://` — GitHub Actions UI helpfully (🙄) removes the protocol when formatting URLs.

**Comment by joesan** [comment]
Do we prepend the URL with https:// or http:// or both and run it through?

**Comment by joesan** [comment]
Seems there is a PR for the Linkify library that is used here: https://github.com/robinst/linkify/issues/7

**Comment by mre** [comment]
Right! Fixing that should fix the problem. The PR is not actively worked on but if you find the time to fix the open issues documented in the issue description then a lot of people would be greatful. 😉

**Comment by joesan** [comment]
I can try that, but for this case here as a quick win, I will submit a PR that checks for E-Mail using Regex so that the cases without a scheme are not treated as E-Mail. Is that ok?

**Comment by joesan** [comment]
Partial fix to not treat Medium links as E-Mail: https://github.com/lycheeverse/lychee/pull/164

**Comment by polyzen** [comment]
This also causes false-negatives for CDN's, eg.:
`https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js (Invalid mail address: https://cdn.jsdelivr.net/npm/jquery@3.5.1/dist/jquery.min.js)`

**Comment by pawroman** [comment]
I'm seeing issues with this too. Pretty much all medium.com links parse as email addresses. Let me see if I can provide a fix :smile:

**Comment by pawroman** [comment]
The PR fixing this properly (hopefully! :crossed_fingers:) is up now.

**Comment by baky0905** [comment]
> In the meantime you could use `--exclude-mail` in case you don't want to check mail addresses anyway. 😅

Hi @mre , I have attempted to use `--exclude-mail           Exclude all mail addresses from checking`  since am not interested in including email, but still it catches the medium articles as emails:

example of the link that was still caught and reported :

↯ https://medium.com/@saswat.sipun/shell-scripting-cheat-sheet-c0ecfb80391 (Invalid mail address: https://medium.com/@saswat.sipun/shell-scripting-cheat-sheet-c0ecfb80391)

**Comment by mre** [comment]
Thanks. Haven't looked at it yet, but my suspicion is that it's because we do the parsing before the rejection of mail addresses in general.

**Comment by pawroman** [comment]
> example of the link that was still caught and reported :
> 
> ↯ https://medium.com/@saswat.sipun/shell-scripting-cheat-sheet-c0ecfb80391 (Invalid mail address: https://medium.com/@saswat.sipun/shell-scripting-cheat-sheet-c0ecfb80391)

The issue should be fixed by #177.

>  Thanks. Haven't looked at it yet, but my suspicion is that it's because we do the parsing before the rejection of mail addresses in general.

#177 makes it so that links are only parsed as emails if they fail parsing as an URL.

**Comment by mre** [comment]
This was fixed by @pawroman, who delivered an amazing PR. Should now work in `master` and `lychee-action@master`. 🥳
